fix issue 1538

evgaster opened the pull request.

evgaster committed:
fix issue 1538

evgaster commented on doc/commands/system.md:
```diff
@@ -161,7 +167,9 @@ This displays on the stack as `HH:MM:SS`.
-Return the Julian day number for the given date and time
+Return the Julian day number for the given date and time.
+
+For dates the Gregorian calendar is assumed. The Gregorian calendar jumps from 1582-10-04 to 1582-10-15. This command ignores that gap. So the Julian day number given by this command for dates on or before 1582-10-14 may devite from other converters. 
```
devite --> deviate

evgaster committed:
fix issue 1538; type corrected

c3d commented:
Hi @evgaster, thank you for all your commits. I am presently busy with a personal matter that slows me down in integrating these changes, but they are very welcome. Keep going!

In order for me to be able to integrate your changes more easily, could you try to follow the standard format for commit messages:

1. First line that begins with the impacted subsystem and describes the commit, e.g. `doc: Improve JulianDayNumber description, and fits in less than 80 characters

2. A description of the commit with additional details, e.g. `The JulianDayNumber deviates from other converters like <URL of alternate converter> in that it takes a Gregorian date as input. Mention it in the documentation.`

3. The issue being fixed with the following format: `Fixes: #1538`.

Having the # sign in front is important, as it makes the issue directly clickable in GitHub. The `Fixes: ` format is also important for automation. GitHub uses it when a new release gets out to close the issues. If the format is wrong the connection between the commit and the issue will not happen.

Thanks again for all your work.

evgaster committed:
doc: Improve the description of "Julian day number" commands

Describe how the commands are linked to the Gregorian calendar.
Refer to an authoritative converter on the world wide web.

Added a description for the DateMenu and TimeMenu commands.
That avoids a "No help for ..." error when (long) holding the key for those.

Signed-off-by: Ed@vanGasteren.net

evgaster committed:
src: Allow for "Julian day number"s with a negative Gregorian year

The function julian_day_number(...) in datetime.cc handled a calculation with the month incorrectly which leads to a wrong "Julian day number". The intermediate value rm is now an int (i.s.o. uint). Months are now handled correctly.

The function date_from_julian_day(...) in datetime.cc did not allow for a negative year which leads to a wrong year for smaller "Julian day number"s. Year is now an int (i.s.o. uint). The sign of a negative year is preserved and returned.

Added comments to both functions about the (most probable) source (references with URLs) of the algorithms used and their working. That took some reverse engineering. Pitty the original author didn't include more information.

Fixes: #1538

Other functions with dates may not handle negative years correctly. That might be(come) an issue but not part of this one.

In test.cc added 2 tests for conversion between "Julian day number" and (Gregorian) date.

Signed-off-by: Ed@vanGasteren.net

evgaster commented:
@c3d I am struggling how to best do this (inline with your commit instructions).

What I did now was added 2 separate commits (I am making changes to 2 subsystems: doc; src) to this pull request. To get there I had to make additional changes to the files.

The new commits have a description which in fact cover the previous commits in this pull request.

If you want me to handle it differently, then please give clear instructions.
